File: sql-insight-engine/sql-insight-engine-innodb/src/main/java/org/gongxuanzhang/sql/insight/core/engine/storage/innodb/page/InnoDbPage.java
/*
 * Copyright 2023 java-mysql  and the original author or authors <gongxuanzhangmelt@gmail.com>.
 *
 * Licensed under the GNU Affero General Public License v3.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://github.com/implement-study/sql-insight/blob/main/LICENSE
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gongxuanzhang.sql.insight.core.engine.storage.innodb.page;


import kotlin.Pair;
import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import org.gongxuanzhang.easybyte.core.ByteWrapper;
import org.gongxuanzhang.easybyte.core.DynamicByteBuffer;
import org.gongxuanzhang.sql.insight.core.engine.storage.innodb.factory.PageFactory;
import org.gongxuanzhang.sql.insight.core.engine.storage.innodb.factory.PageHeaderFactory;
import org.gongxuanzhang.sql.insight.core.engine.storage.innodb.index.InnodbIndex;
import org.gongxuanzhang.sql.insight.core.engine.storage.innodb.page.compact.IndexRecord;
import org.gongxuanzhang.sql.insight.core.engine.storage.innodb.page.compact.RecordHeader;
import org.gongxuanzhang.sql.insight.core.engine.storage.innodb.page.compact.RecordType;
import org.gongxuanzhang.sql.insight.core.engine.storage.innodb.utils.PageSupport;
import org.gongxuanzhang.sql.insight.core.exception.DuplicationPrimaryKeyException;
import org.jetbrains.annotations.NotNull;

import java.util.Comparator;
import java.util.Iterator;
import java.util.NoSuchElementException;

import static org.gongxuanzhang.sql.insight.core.engine.storage.innodb.page.Constant.SLOT_MAX_COUNT;
import static org.gongxuanzhang.sql.insight.core.engine.storage.innodb.page.ConstantSize.FILE_HEADER;
import static org.gongxuanzhang.sql.insight.core.engine.storage.innodb.page.ConstantSize.FILE_TRAILER;
import static org.gongxuanzhang.sql.insight.core.engine.storage.innodb.page.ConstantSize.INFIMUM;
import static org.gongxuanzhang.sql.insight.core.engine.storage.innodb.page.ConstantSize.PAGE;
import static org.gongxuanzhang.sql.insight.core.engine.storage.innodb.page.ConstantSize.PAGE_HEADER;
import static org.gongxuanzhang.sql.insight.core.engine.storage.innodb.page.ConstantSize.SUPREMUM;

/**
 * InnoDb Page
 * size default 16K.
 *
 * @author gxz gongxuanzhangmelt@gmail.com
 **/

@Data
@Slf4j
public abstract class InnoDbPage implements ByteWrapper, Comparator<InnodbUserRecord>, PageObject,
        Iterable<InnodbUserRecord> {


    /**
     * 38 bytes
     **/
    FileHeader fileHeader;
    /**
     * 56 bytes
     **/
    PageHeader pageHeader;
    /**
     * 13 bytes
     **/
    Infimum infimum;
    /**
     * 13 bytes
     **/
    Supremum supremum;
    /**
     * uncertain bytes.
     * user records bytes + freeSpace = page size - other fixed size
     **/
    UserRecords userRecords;
    /**
     * uncertain bytes.
     **/
    PageDirectory pageDirectory;
    /**
     * 8 bytes
     **/
    FileTrailer fileTrailer;

    PageExt ext;

    protected InnoDbPage(InnodbIndex index) {
        this.ext = new PageExt();
        this.ext.belongIndex = index;
    }

    @Override
    public byte[] toBytes() {
        DynamicByteBuffer buffer = DynamicByteBuffer.allocate();
        buffer.append(fileHeader.toBytes());
        buffer.append(pageHeader.toBytes());
        buffer.append(infimum.toBytes());
        buffer.append(supremum.toBytes());
        buffer.append(userRecords.toBytes());
        buffer.append(new byte[getFreeSpace()]);
        buffer.append(pageDirectory.toBytes());
        buffer.append(fileTrailer.toBytes());
        return buffer.toBytes();
    }


    public PageType pageType() {
        return PageType.valueOf(this.fileHeader.pageType);
    }

    public short getFreeSpace() {
        return (short) (PAGE.size() -
                PAGE_HEADER.size() -
                FILE_HEADER.size() -
                FILE_TRAILER.size() -
                SUPREMUM.size() -
                INFIMUM.size() -
                this.pageDirectory.length() - userRecords.length());
    }

    /**
     * add a data.
     * data means a insert row .
     * page is leaf node will insert data.
     * page is index node will find target leaf node and insert data.
     * may be split page in process
     **/
    public void insertData(InnodbUserRecord data) {
        Pair<InnodbUserRecord, InnodbUserRecord> preAndNext = findPreAndNext(data);
        InnodbUserRecord pre = preAndNext.getFirst();
        InnodbUserRecord next = preAndNext.getSecond();
        linkedAndAdjust(pre, data, next);
        splitIfNecessary();
        PageSupport.flushPage(this);
    }


    /**
     * find location where the record linked in page and return pre and next.
     * search only,don't affect the page .
     *
     * @param userRecord target record
     **/
    protected Pair<InnodbUserRecord, InnodbUserRecord> findPreAndNext(InnodbUserRecord userRecord) {
        int targetSlot = findTargetSlot(userRecord);
        InnodbUserRecord pre = getUserRecordByOffset(pageDirectory.indexSlot(targetSlot + 1));
        InnodbUserRecord next = getUserRecordByOffset(pre.nextRecordOffset() + pre.offset());
        while (this.compare(userRecord, next) > 0) {
            pre = next;
            next = getUserRecordByOffset(pre.nextRecordOffset() + pre.offset());
        }
        return new Pair<>(pre, next);
    }


    /**
     * find the slot where the target record is located
     * <p>
     * return 0 means supremum
     *
     * @return slot index is user record inserted never return {@code slot.length -1 } because slot.length - 1 is the
     * infimum
     **/
    protected int findTargetSlot(InnodbUserRecord userRecord) {
        int left = 0;
        int right = pageDirectory.slotCount() - 1;
        while (left < right - 1) {
            int mid = (right + left) / 2;
            short offset = this.pageDirectory.getSlots()[mid];
            InnodbUserRecord base = getUserRecordByOffset(offset);
            int compare = this.compare(userRecord, base);
            if (compare == 0) {
                throw new DuplicationPrimaryKeyException(base.getRowId());
            }
            if (compare < 0) {
                left = mid;
            } else {
                right = mid;
            }
        }
        InnodbUserRecord base = getUserRecordByOffset(this.pageDirectory.getSlots()[right]);
        int compare = this.compare(userRecord, base);
        if (compare == 0) {
            throw new DuplicationPrimaryKeyException(base.getRowId());
        }
        if (compare > 0) {
            return left;
        }
        return right;
    }

    /**
     * @param offsetInPage offset in page
     * @return user record
     **/
    protected InnodbUserRecord getUserRecordByOffset(int offsetInPage) {
        if (offsetInPage == ConstantSize.INFIMUM.offset()) {
            return this.infimum;
        }
        if (offsetInPage == ConstantSize.SUPREMUM.offset()) {
            return this.supremum;
        }
        InnodbUserRecord wrap = wrapUserRecord(offsetInPage);
        wrap.setOffset(offsetInPage);
        return wrap;
    }


    protected abstract InnodbUserRecord wrapUserRecord(int offsetInPage);


    /**
     * this page should split.
     * in general after insert row call this method
     **/
    protected abstract void splitIfNecessary();


    protected void upgrade(InnoDbPage prePage, InnoDbPage nextPage) {
        //  is root page
        if (this.ext.parent == null) {
            rootPageUpgrade(prePage, nextPage);
        } else {
            // normal leaf node
            prePage.fileHeader.setOffset(this.fileHeader.offset);
            int newDataPageOffset = PageSupport.allocatePage(this.ext.belongIndex);
            nextPage.fileHeader.setOffset(newDataPageOffset);
            InnoDbPage parent = this.ext.parent;
            this.transferFrom(prePage);
            parent.insertData(nextPage.pageIndex());
        }
    }


    /**
     * get the first index node to parent node insert
     **/
    public abstract IndexRecord pageIndex();


    /**
     * root page upgrade.
     * create 2 new page and linked whether upgrading  from index page or data page.
     **/
    private void rootPageUpgrade(InnoDbPage preChild, InnoDbPage secondChild) {
        preChild.pageHeader.level = this.pageHeader.level;
        secondChild.pageHeader.level = this.pageHeader.level;
        this.pageHeader.level++;
        FileHeader firstFileHeader = preChild.getFileHeader();
        FileHeader secondFileHeader = secondChild.getFileHeader();
        int offset = PageSupport.allocatePage(this.ext.belongIndex, 2);
        firstFileHeader.setOffset(offset);
        secondFileHeader.setOffset(offset + ConstantSize.PAGE.size());
        firstFileHeader.setPre(-1);
        firstFileHeader.setNext(secondFileHeader.offset);
        secondFileHeader.setPre(firstFileHeader.offset);
        secondFileHeader.setNext(-1);
        //  transfer to index page
        this.fileHeader.next = -1;
        this.fileHeader.pageType = PageType.FIL_PAGE_INODE.getValue();
        this.pageHeader = PageHeaderFactory.createPageHeader();
        this.pageDirectory = new PageDirectory();
        //  clear user record
        this.userRecords = new UserRecords();
        this.insertData(preChild.pageIndex());
        this.insertData(secondChild.pageIndex());
    }


    protected void linkedAndAdjust(InnodbUserRecord pre, InnodbUserRecord insertRecord, InnodbUserRecord next) {
        RecordHeader insertHeader = insertRecord.getRecordHeader();
        insertHeader.setHeapNo(this.pageHeader.absoluteRecordCount);
        insertRecord.setOffset(this.pageHeader.lastInsertOffset + insertRecord.beforeSplitOffset());
        insertHeader.setNextRecordOffset(next.offset() - insertRecord.offset());
        pre.getRecordHeader().setNextRecordOffset(insertRecord.offset() - pre.offset());
        refreshRecordHeader(pre);

        insertHeader.setRecordType(RecordType.NORMAL);

        //  adjust page
        this.userRecords.addRecord(insertRecord);
        this.pageHeader.absoluteRecordCount++;
        this.pageHeader.recordCount++;
        this.pageHeader.heapTop += (short) insertRecord.length();
        this.pageHeader.lastInsertOffset += (short) insertRecord.length();


        //  adjust group
        while (next.getRecordHeader().getNOwned() == 0) {
            next = getUserRecordByOffset(next.offset() + next.nextRecordOffset());
        }
        RecordHeader nextHeader = next.getRecordHeader();
        int groupCount = nextHeader.getNOwned();
        nextHeader.setNOwned(groupCount + 1);
        if (next.getRecordHeader().getNOwned() <= SLOT_MAX_COUNT) {
            return;
        }
        log.info("start group split ...");
        for (int i = 0; i < this.pageDirectory.slots.length - 1; i++) {
            if (this.pageDirectory.slots[i] == next.offset()) {
                InnodbUserRecord preGroupMax = getUserRecordByOffset(this.pageDirectory.slots[i + 1]);
                for (int j = 0; j < SLOT_MAX_COUNT >> 1; j++) {
                    preGroupMax = getUserRecordByOffset(preGroupMax.offset() + preGroupMax.nextRecordOffset());
                }
                this.pageDirectory.split(i, (short) preGroupMax.offset());
            }
        }
        log.info("end group split ...");
    }


    /**
     * byte array copy from target page
     **/
    public void transferFrom(InnoDbPage page) {
        InnoDbPage snapshot = PageFactory.swap(page.toBytes(), this.ext.belongIndex);
        this.fileHeader = snapshot.fileHeader;
        this.pageHeader = snapshot.pageHeader;
        this.infimum = snapshot.infimum;
        this.supremum = snapshot.supremum;
        this.userRecords = snapshot.userRecords;
        this.pageDirectory = snapshot.pageDirectory;
        this.fileTrailer = snapshot.fileTrailer;
        this.ext = snapshot.ext;
    }

    private void refreshRecordHeader(InnodbUserRecord record) {
        if(record instanceof Infimum){
            return;
        }
        if(record instanceof Supremum){
            return;
        }
        byte[] headerByte = record.getRecordHeader().toBytes();
        int bodyOffset = record.offset() - ConstantSize.RECORD_HEADER.size() - PAGE_HEADER.size() -
                FILE_HEADER.size() -
                SUPREMUM.size() -
                INFIMUM.size();
        System.arraycopy(headerByte, 0, this.userRecords.body, bodyOffset, headerByte.length);
    }

    @Override
    public int length() {
        return ConstantSize.PAGE.size();
    }

    @NotNull
    @Override
    public Iterator<InnodbUserRecord> iterator() {
        return new Itr();
    }

    public class Itr implements Iterator<InnodbUserRecord> {
        InnodbUserRecord cursor = getUserRecordByOffset(infimum.nextRecordOffset() + infimum.offset());

        @Override
        public boolean hasNext() {
            return cursor != InnoDbPage.this.supremum;
        }

        @Override
        public InnodbUserRecord next() {
            if (cursor == supremum) {
                throw new NoSuchElementException();
            }
            InnodbUserRecord result = cursor;
            cursor = getUserRecordByOffset(cursor.nextRecordOffset() + cursor.offset());
            return result;
        }
    }

    @Data
    public static class PageExt {
        InnodbIndex belongIndex;
        IndexPage parent;
    }

}

File: sql-insight-engine/sql-insight-engine-innodb/src/main/kotlin/tech/insight/engine/innodb/page/InnoDbPage.kt
package tech.insight.engine.innodb.page

import io.netty.buffer.ByteBuf
import tech.insight.core.annotation.Temporary
import tech.insight.core.bean.condition.Expression
import tech.insight.core.logging.Logging
import tech.insight.engine.innodb.core.InnodbSessionContext
import tech.insight.engine.innodb.core.buffer.BufferPool
import tech.insight.engine.innodb.index.InnodbIndex
import tech.insight.engine.innodb.page.compact.Compact
import tech.insight.engine.innodb.page.compact.RecordHeader
import tech.insight.engine.innodb.page.compact.RowFormatFactory
import tech.insight.engine.innodb.page.type.PageType


/**
 * InnoDb Page
 * source is 16K bytes
 * @author gxz gongxuanzhangmelt@gmail.com
 */
class InnoDbPage(internal val source: ByteBuf, index: InnodbIndex) : Logging(), PageObject,
    Iterable<InnodbUserRecord>, Comparator<InnodbUserRecord> {

    init {
        require(source.capacity() == ConstantSize.PAGE.size) {
            "page size must be 16K"
        }
    }

    /**
     * 38 bytes
     */
    val fileHeader = FileHeader(this)

    /**
     * 56 bytes
     */
    val pageHeader = PageHeader(this)

    /**
     * 13 bytes
     */
    val infimum = Infimum(this)

    /**
     * 13 bytes
     */
    val supremum = Supremum(this)

    /**
     * uncertain bytes.
     * user records bytes + freeSpace = page size - other fixed size
     */
    val userRecords = UserRecords(this)

    /**
     * uncertain bytes.
     */
    val pageDirectory = PageDirectory(this)

    /**
     * 8 bytes
     */
    val fileTrailer = FileTrailer(this)

    /**
     * some extra info for page
     */
    var ext: PageExt

    init {
        ext = PageExt()
        ext.belongIndex = index
    }

    override fun toBytes(): ByteArray {
        return source.array()
    }

    /**
     * the page free space length
     */
    fun remainSpace() = ConstantSize.FILE_TRAILER.offset - pageHeader.slotCount * Short.SIZE_BYTES - pageHeader.heapTop

    fun pageType(): PageType {
        val currentType = ext.pageType
        if (currentType == null || currentType.value != fileHeader.pageType) {
            ext.pageType = PageType.valueOf(fileHeader.pageType, this)
        }
        return ext.pageType!!
    }

    val freeSpace: UShort
        get() = (ConstantSize.PAGE.size -
                ConstantSize.PAGE_HEADER.size -
                ConstantSize.FILE_HEADER.size -
                ConstantSize.FILE_TRAILER.size -
                ConstantSize.SUPREMUM.size -
                ConstantSize.INFIMUM.size -
                pageDirectory.length() - userRecords.length()).toUShort()

    /**
     * add a data.
     * data means a insert row .
     * page is leaf node will insert data.
     * page is index node will find target leaf node and insert data.
     * may be split page in process
     */
    fun insertData(data: InnodbUserRecord) {
        val targetPage = this.locatePage(data)
        targetPage.doInsertData(data)
    }

    /**
     * link [PageType.locatePage]
     */
    fun locatePage(data: InnodbUserRecord): InnoDbPage {
        return this.pageType().locatePage(data)
    }


    /**
     *
     */
    private fun doInsertData(data: InnodbUserRecord) {
        InnodbSessionContext.getInnodbSessionContext().modifyPage(this)
        var preCandidate = pageDirectory.findTargetIn(data).minRecord()
        var nextCandidate = preCandidate.nextRecord()
        //   <= ? <  this data already existed
        while (nextCandidate <= data) {
            preCandidate = nextCandidate
            nextCandidate = preCandidate.nextRecord()
        }
        val preFree = this.freeSpace
        linkedAndAdjust(preCandidate, data, nextCandidate)
        val after = this.freeSpace
        val diff = preFree - after
        debug { "data: ${data.toBytes().size} diff: $diff pre: $preFree after: $after " }
        pageSplitIfNecessary()
    }


    /**
     * data page will split when free space less than one in thirty-two page size
     */
    private fun pageSplitIfNecessary() {
        if (freeSpace.toInt() > ConstantSize.PAGE.size shr 4) {
            return
        }
        val pageUserRecord: MutableList<InnodbUserRecord> = ArrayList(pageHeader.recordCount + 1)
        var base: InnodbUserRecord = infimum
        var allLength = 0
        while (true) {
            base = base.nextRecord()
            if (base === supremum) {
                break
            }
            pageUserRecord.add(base)
            allLength += base.length()
        }
        when (computeSplitStrategy()) {
            PageSplitStrategy.MIDDLE_SPLIT -> middleSplit(pageUserRecord, allLength)
            else -> TODO("other split strategy not implement")
        }
    }


    /**
     * middle split.
     * insert direction unidentified (directionCount less than 5)
     *
     *
     * if this page is root page.
     * transfer root page to index page from data page.
     * create two data page linked.
     * if this page is normal leaf node,
     * create a data page append to index file and insert a index record to parent (index page)
     *
     * @param pageUserRecord all user record in page with inserted
     * @param allLength      all user record length
     */
    private fun middleSplit(pageUserRecord: List<InnodbUserRecord>, allLength: Int) {
        var remainLength = allLength
        val half = remainLength / 2
        val middleIndex = pageUserRecord.indexOfFirst {
            remainLength -= it.length()
            remainLength <= half
        }
        if (ext.parent == null) {
            val (leftPage, rightPage) = splitToSubPage(pageUserRecord, middleIndex, this)
            return this.pageType().rootUpgrade(leftPage, rightPage)
        }
        val remainRecords = pageUserRecord.subList(0, middleIndex)
        coverRecords(remainRecords)
        val otherPage = BufferPool.allocatePage(this.ext.belongIndex)
        val otherRecords = pageUserRecord.subList(middleIndex, pageUserRecord.size)
        otherPage.coverRecords(otherRecords)
        return this.pageType().upgrade(otherPage)
    }


    /**
     * page split create a new page.
     * page only have user records
     * @param recordList data or index in the page that sorted
     * @return first of pair is left node,second is right node
     */
    private fun splitToSubPage(
        recordList: List<InnodbUserRecord>,
        splitIndex: Int,
        parentPage: InnoDbPage
    ): Pair<InnoDbPage, InnoDbPage> {
        val index = parentPage.ext.belongIndex
        val leftRecords = recordList.subList(0, splitIndex)
        val rightRecords = recordList.subList(splitIndex, recordList.size)
        val leftPage = BufferPool.allocatePage(index)
        leftPage.coverRecords(leftRecords)
        val rightPage = BufferPool.allocatePage(index)
        rightPage.coverRecords(rightRecords)
        return Pair(leftPage, rightPage)
    }

    /**
     * At present there is only a middle split
     */
    private fun computeSplitStrategy(): PageSplitStrategy {
        if (pageHeader.directionCount < Constant.DIRECTION_COUNT_THRESHOLD) {
            return PageSplitStrategy.MIDDLE_SPLIT
        }
        return PageSplitStrategy.NOT_SPLIT
    }

    /**
     * @param offsetInPage offset in page, offset is after record header , aka    |vars|null|record header| here!| record body|
     * @return user record
     */
    fun getUserRecordByOffset(offsetInPage: Int): InnodbUserRecord {
        if (offsetInPage == this.infimum.absoluteOffset()) {
            return infimum
        }
        if (offsetInPage == this.supremum.absoluteOffset()) {
            return supremum
        }
        return this.pageType().convertUserRecord(offsetInPage)
    }


    /**
     * get the first user record in page user records linked.
     */
    fun getFirstUserRecord(): InnodbUserRecord {
        if (this.pageHeader.recordCount == 0) {
            throw NoSuchElementException("page is empty")
        }
        return getUserRecordByOffset(infimum.absoluteOffset() + infimum.nextRecordOffset())
    }

    /**
     * @param offset record offset, the record header offset = offset - record header size
     * @return record
     */
    fun readRecordHeader(offset: Int): RecordHeader {
        val recordHeaderSize: Int = ConstantSize.RECORD_HEADER.size
        return RecordHeader(source.slice(offset - recordHeaderSize, recordHeaderSize))
    }

    /**
     * get the first index node to parent node insert
     * link [PageType.pageIndex]
     */
    fun pageIndex(): InnodbUserRecord {
        return pageType().pageIndex()
    }


    fun update(oldRow: Compact, updateFields: Map<String, Expression>) {
        InnodbSessionContext.getInnodbSessionContext().modifyPage(this)
        val updateCompact = RowFormatFactory.compactFromUpdateRow(oldRow, updateFields)
        return replace(oldRow, updateCompact)
    }

    private fun linkedAndAdjust(pre: InnodbUserRecord, insertRecord: InnodbUserRecord, next: InnodbUserRecord) {
        val recordInPage = userRecords.addRecord(insertRecord)
        pre.linkRecord(recordInPage)
        recordInPage.linkRecord(next)
        val groupMax = next.groupMax()
        if (++groupMax.recordHeader.nOwned <= Constant.SLOT_MAX_COUNT) {
            return
        }
        debug { "occurred group split ..." }
        val splitSlot = pageDirectory.requireSlotByOffset(groupMax.absoluteOffset())
        //   todo group split strategy
        val leftGroupCount = Constant.SLOT_MAX_COUNT shr 1
        val leftMaxRecord = run {
            var candidate = splitSlot.smaller().maxRecord()
            repeat(leftGroupCount) {
                candidate = candidate.nextRecord()
            }
            candidate.apply {
                recordHeader.nOwned = leftGroupCount
            }
        }
        groupMax.recordHeader.nOwned = Constant.SLOT_MAX_COUNT - leftGroupCount + 1
        pageDirectory.insert(splitSlot.index, leftMaxRecord.absoluteOffset())
    }


    fun coverRecords(records: List<InnodbUserRecord>) {
        this.clear()
        records.forEach { this.insertData(it) }
        //   todo batch insert 
    }

    /**
     * remove all user records in this page,However, it does not change the position of the page in the whole b+ tree.
     * in other words file header and file trailer is not changed.
     */
    internal fun clear() {
        this.userRecords.clear()
    }

    override fun length(): Int {
        return ConstantSize.PAGE.size
    }

    override val belongPage: InnoDbPage = this

    override fun iterator(): Iterator<InnodbUserRecord> {
        return Itr()
    }

    override fun hashCode(): Int {
        return fileHeader.spaceId + 31 * fileHeader.offset
    }

    override fun compare(o1: InnodbUserRecord, o2: InnodbUserRecord): Int {
        return this.pageType().compare(o1, o2)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is InnoDbPage) return false

        if (fileHeader != other.fileHeader) return false
        return true
    }


    inner class Itr : Iterator<InnodbUserRecord> {
        private var cursor = getUserRecordByOffset(infimum.nextRecordOffset() + infimum.absoluteOffset())
        override fun hasNext(): Boolean {
            return cursor !== supremum
        }

        override fun next(): InnodbUserRecord {
            if (cursor === supremum) {
                throw NoSuchElementException()
            }
            val result = cursor
            cursor = getUserRecordByOffset(cursor.nextRecordOffset() + cursor.absoluteOffset())
            return result
        }

    }

    inner class PageExt {

        lateinit var belongIndex: InnodbIndex

        var parent: InnoDbPage? = null

        var pageType: PageType? = null

        var change: Boolean = true

        var bytes: ByteArray = ByteArray(0)


        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is PageExt) return false

            if (belongIndex != other.belongIndex) return false
            if (parent != other.parent) return false

            return true
        }

        override fun hashCode(): Int {
            var result = belongIndex.hashCode()
            result = 31 * result + (parent?.hashCode() ?: 0)
            return result
        }

    }


    companion object {


        @Temporary("direct get from buffer pool")
        fun findPageByOffset(pageOffset: Int, index: InnodbIndex): InnoDbPage {
            return BufferPool.getPageAndCache(pageOffset, index)
        }


    }
}
